src/lib/yang/binary.lua

```diff
@@ -244,6 +244,11 @@ local function data_emitter(production)
             stream:write_stringref('stringref')
             stream:write_stringref(data)
          end
+      elseif primitive_type == 'empty' then
+         return function (data, stream)
+            stream:write_stringref('stringref')
+            stream:write_stringref('')
```

Comment:
I agree the original code was not nice and now I understand better how things work. However I think the original code still would make the job. The reason is that in the end what it does is to mark the presence of a leaf of data type empty with an empty string. If the leaf was not present simply the compilation code is not executed. I don't see how it would be possible that during deserialization phase there was a value '' if an empty node was not present.

OTOH, I think value of `data` is always true. The reason is that empty nodes get parsed as:

```lua
types.empty = {}
function types.empty.parse (str, what)
   return assert(str == nil, "not empty value for "..what)
end   
```

Considering that I wonder if it would be enough to simply mark the existence of the node. Something like:

```
stream:write_stringref('exists')
```

```lua
function readers.exists()
   return true
end
```

In any case, I agree with the proposed change and having a boolean flag is more generic.